examples/go-app-counter/main.go

```diff
@@ -0,0 +1,324 @@
+						Bytecode:    paramData.BytecodeSource,
+						Name:        "uprobe_counter",
+						ProgramType: *bpfmanHelpers.Kprobe.Uint32(),
+						Attach: &gobpfman.AttachInfo{
+							Info: &gobpfman.AttachInfo_UprobeAttachInfo{
+								UprobeAttachInfo: &gobpfman.UprobeAttachInfo{
+									FnName: &fnName,
+									Target: "libc",
+								},
+							},
+						},
+						MapOwnerId: &mapOwnerId,
+					}
+				} else {
+					loadRequest = &gobpfman.LoadRequest{
+						Bytecode:    paramData.BytecodeSource,
+						Name:        "uprobe_counter",
+						ProgramType: *bpfmanHelpers.Kprobe.Uint32(),
+						Attach: &gobpfman.AttachInfo{
+							Info: &gobpfman.AttachInfo_UprobeAttachInfo{
+								UprobeAttachInfo: &gobpfman.UprobeAttachInfo{
+									FnName: &fnName,
+									Target: "libc",
+								},
+							},
+						},
+					}
+				}
+			case configMgmt.ProgTypeXdp:
+				if paramData.MapOwnerId != 0 {
+					mapOwnerId := uint32(paramData.MapOwnerId)
+					loadRequest = &gobpfman.LoadRequest{
+						Bytecode:    paramData.BytecodeSource,
+						Name:        "xdp_stats",
+						ProgramType: *bpfmanHelpers.Xdp.Uint32(),
+						Attach: &gobpfman.AttachInfo{
+							Info: &gobpfman.AttachInfo_XdpAttachInfo{
+								XdpAttachInfo: &gobpfman.XDPAttachInfo{
+									Priority: int32(paramData.Priority),
+									Iface:    paramData.Iface,
+								},
+							},
+						},
+						MapOwnerId: &mapOwnerId,
+					}
+				} else {
+					loadRequest = &gobpfman.LoadRequest{
+						Bytecode:    paramData.BytecodeSource,
+						Name:        "xdp_stats",
+						ProgramType: *bpfmanHelpers.Xdp.Uint32(),
+						Attach: &gobpfman.AttachInfo{
+							Info: &gobpfman.AttachInfo_XdpAttachInfo{
+								XdpAttachInfo: &gobpfman.XDPAttachInfo{
+									Priority: int32(paramData.Priority),
+									Iface:    paramData.Iface,
+								},
+							},
+						},
+					}
+				}
+			}
+
+			// 1. Load Program using bpfman
+			var res *gobpfman.LoadResponse
+			res, err = c.Load(ctx, loadRequest)
+			if err != nil {
+				conn.Close()
+				log.Print(err)
+				return nil, action, err
+			}
+
+			kernelInfo := res.GetKernelInfo()
+			if kernelInfo != nil {
+				paramData.ProgId = uint(kernelInfo.GetId())
+			} else {
+				conn.Close()
+				log.Printf("kernelInfo not returned in LoadResponse")
+				return nil, action, err
+			}
+			log.Printf("Program registered with id %d\n", paramData.ProgId)
+
+			// 2. Set up defer to unload program when this is closed
+			defer func(id uint) {
+				log.Printf("unloading program: %d\n", id)
+				_, err = c.Unload(ctx, &gobpfman.UnloadRequest{Id: uint32(id)})
+				if err != nil {
+					conn.Close()
+					log.Print(err)
+					return
+				}
+				conn.Close()
+			}(paramData.ProgId)
+
+			// 3. Get access to our map
+			mapPath, err = configMgmt.CalcMapPinPath(res.GetInfo(), mapIndex)
+			if err != nil {
+				log.Print(err)
+				return nil, action, err
+			}
+		} else {
+			// 2. Set up defer to close connection
+			defer func(id uint) {
```

Comment:
I think this deferred connection close will be a problem too.

Reply:
yeah I agree for the above two comments the defer needs to move up one level not in the common function I think